rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -441,30 +461,111 @@ BtcBlockStoreWithCache getBtcBlockStore() {
+                return new PeginEvaluationResult(PeginProcessAction.CANNOT_BE_PROCESSED, PEGIN_V1_INVALID_PAYLOAD);
+            }
+        }
 
-        return PegUtilsLegacy.getTransactionType(
-            btcTx,
-            getActiveFederation(),
-            getRetiringFederation(),
-            retiredFederationP2SHScript,
-            oldFederationAddress,
-            activations,
-            minimumPeginTxValue,
-            federationWallet
+        if (peginInformation.getProtocolVersion() == 0) {
+            switch (peginInformation.getSenderBtcAddressType()) {
+                case P2PKH:
+                case P2SHP2WPKH:
+                    return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REGISTERED);
+                case P2SHMULTISIG:
+                case P2SHP2WSH:
+                    return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REFUNDED, LEGACY_PEGIN_MULTISIG_SENDER);
+                default:
+                    return new PeginEvaluationResult(PeginProcessAction.CANNOT_BE_PROCESSED, LEGACY_PEGIN_UNDETERMINED_SENDER);
+            }
+        } else if (peginInformation.getProtocolVersion() == 1) {
+            return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REGISTERED);
+        }
+        throw new IllegalStateException("Pegin could not be evaluated.");
+    }
+
+    protected void processPegIn(BtcTransaction btcTx, Transaction rskTx, int height, Sha256Hash btcTxHash) throws IOException, RegisterBtcTransactionException {
+        if (activations.isActive(ConsensusRule.RSKIP379)){
+            tryToProcessPegin(btcTx, rskTx);
+        } else {
+            legacyProcessPegin(btcTx, rskTx, height, btcTxHash);
+        }
+    }
+
+    private void tryToProcessPegin(BtcTransaction btcTx, Transaction rskTx) throws IOException, RegisterBtcTransactionException {
+        Coin totalAmount = computeTotalAmountSent(btcTx);
+
+        PeginInformation peginInformation = new PeginInformation(
+            btcLockSenderProvider,
+            peginInstructionsProvider,
+            activations
         );
+
+        PeginEvaluationResult peginEvaluationResult = evaluatePegin(btcTx, peginInformation);
+
+        Optional<RejectedPeginReason> rejectedPeginReason = peginEvaluationResult.getRejectedPeginReason();
+        switch (peginEvaluationResult.getPeginProcessAction()) {
+            case CAN_BE_REGISTERED:
+                executePegIn(btcTx, peginInformation, totalAmount);
+                markTxAsProcessed(btcTx);
+                break;
+            case CAN_BE_REFUNDED:
+                rejectedPeginReason.ifPresent(rejectedReason -> eventLogger.logRejectedPegin(btcTx, rejectedReason));
+                generateRejectionRelease(btcTx, peginInformation.getBtcRefundAddress(), rskTx, totalAmount);
+                markTxAsProcessed(btcTx);
+                throw new RegisterBtcTransactionException(String.format("Error while trying to parse peg-in information for tx %s", btcTx.getHash()));
```

Comment:
Why throw this exception here? The transaction was refunded, why is an exception? I would simply add a log for each case stating what happened